README
Author: Sergey Nehaenko <sergey.nekhaenko@gmail.com>
Current Version: 1.1
Last Update: 19.12.2012
License: GPL

Demo: http://detector.endorphinua.ru/
Twitter: https://twitter.com/endorphinua


changelog.txt
v 1.1.11
	- add Edge browser detection

v 1.1.10
	- add $isBot property
	- add Bot Detection (Google, Yandex)

v 1.1.9:
	- fix Lumia`s detection

v 1.1.8:
	device support:
		- HTC 6990
		- HTC Mazaa
		- HTC Mozart
		- HTC 7 Pro
		- HTC 7 Trophy
		- Nokia Lumia 505
		- Nokia Lumia 510
		- Nokia Lumia 520
		- Nokia Lumia 610
		- Nokia Lumia 620
		- Nokia Lumia 710
		- Nokia Lumia 720
		- Nokia Lumia 730
		- Nokia Lumia 800
		- Nokia Lumia 810
		- Nokia Lumia 820
		- Nokia Lumia 822
		- Nokia Lumia 910
		- Nokia Lumia 920
		- Nokia Lumia 928
	fixed:
		- win os family detection
		- added detection version of Windows Phone OS
		- fix os['type'] bug for Windows Phone OS
	added:
		- $debug property (default = false)
		- printAll() function
		- getAll() function
	deleted:
		- tests folder
v 1.1.7:
	browsers support:
		- Chimera
	device support:
		- Google Nexus
		- Dell Android phones
		- Asus tablets
		- Cisco Cius
		- HTC Evo 3D
		- Nokia Lumia 900
		- NOOK Color
		- NOOK Tablet
		- NOOK HD
		- NOOK HD+
		- Pandigital Nova Digital Reader
		- Pandigital SuperNova 8
		- Samsung Focus Flash
		- Toshiba Excite 10 T
		- Toshiba Excite 7 T
		- Toshiba Thrive 7
		- Microsoft Surface
	fixed:
		- Motorolla phones detection
		- Sumsung phones detection

v 1.1.6:
	browsers support:
		- Beamrise
		- Camino
		- Columbus
		- Deepnet Explorer
		- CometBird
		- Kylo
		- iCab
v 1.1.5:
	added tests
	browsers support:
		- Skyfire
		- Dolphin
		- AOL Explorer
		- Konqueror
		- IceWeasel
		
v 1.1.4:
	browser support:
		- Columbus
		- Flock
		
v 1.1.3:
	browsers support:
		- Atomic
		- Chrome Mobile
		- Arora
		- Conkeror
		- Epiphany
		- Midori
		- Maxthon
		- Rekonq
		- Yandex Browser
v 1.1.2:
	- fixed Android tablets detection
	- fixed Chrome mobile detection (for Android and iOS)
v 1.1.1:
	- fix mobile Firefox detection
	- fix Palm OS detection
	- update Demo
v 1.1:
	- Code refactoring
v 1.0.6: 
	- fixed detection of iOS OS
	os support:
		- J2ME
	browsers support:
		- Opera Mobile
	device support:
		- Google Nexus One
		- Google Ion
		- Google Nexus S
		- Google Nexus S 4G
		- iPod
		- LG
		- LG Viewty
		- LG Pop
		- LG dLite
		- LG Expo
		- LG Mini
		- LG Crystal
		- LG Eigen
		- LG Layla
v 1.0.5:
	- fixed detection of Nokia devices
	- fixed detection of BlackBerry devices
	- fixed detection of Symbian OS
	- fixed detection of Windows ME OS
	- fixed detection of Windows Phone OS
	
	os support:
		- HP-UX
		- IRIX
		- Joli OS
		- Nintendo Wii
		- Open BSD
		- Mac OS X 10.8 Mountain Lion
		- Palm OS
		- PClinuxOS
		- RIM OS
		- RIM Tablet OS
		- Solaris
		- Syllable
		- Tizen
		- webOS
		- Windows Mobile
		
	device support:
		- Sony Playstation
		- PSP
		
v 1.0.4:
	os support:
		- Symbian
		- Android
		- Bada
		- Chromium OS
		- FreeBSD
		- Arch Linux
		- Debian
		- Fedora
		- Gentoo
		- Kanotix
		- Mandriva
		- Mint
		- Red Hat
		- Slackware
		- SUSE
		- Max OS X
		- Mac OS X 10.3 Panther
		- Mac OS X 10.4 Tiger
		- Mac OS X 10.5 Leopard
		- Mac OS X 10.6 Snow Leopard
		- Mac OS X 10.7 Lion
		- Morph OS
		- Minix 3
		- NetBSD
		- AROS
		- Amiga OS
		- BeOS
		- DragonFly BSD

v 1.0.3
	os support:
		- iOS
		- Linux
		- Ubuntu
		- Windows 95
		- Windows 98
		- Windows 2000
		- Windows ME
		- Windows CE
		- Windows Server 2003
		- Windows XP
		- Windows Vista
		- Windows 7
		- Windows 8
		- Windows Phone	
v 1.0.2:
	device support:
		- Motorola
		- BenQ
v 1.0.1:
	browser support:
		- Mozilla Firefox
		- Google Chrome
		- Chromium
		- Opera
		- Opera Mini
		- Aurora
		- Dooble
		- Midori
		- Reconq
		- Epiphany
		- Safari
		- Safari Mobile
		- Skyfire
		- ABrowse
		- Acoo Browser
		- America Online Browser
		- Amiga Voyager
		- AOL
		- Avant Browser
		- Beonex
		- BonEcho
		- Browzar
		- Camino
		- Cheshire
		- Chimera
		- ChromePlus
		- Classilla
		- CometBird
		- Comodo Dragon
		- Conkeror
		- Crazy Browser
		- Cyberdog
		- DeskBrowse
		- Deepnet Explorer
	device support:
		- iPhone
		- iPad
		- Amazon Kindle
		- BlackBerry
		- Nokia
		- HTC Desire
		- HTC Dream
		- HTC Flyer
		- HTC Wildfire
		- HTC Desire HD SoftBank
		- HTC Faraday
		- HTC Excalibur
		- HTC One S
		- HTC One V
		- HTC Desire S
		- HTC Desire Z
		- HTC Desire HD
		- HTC Espresso
		- HTC Fiesta
		- HTC Fuwa
		- Siemens
		- Philips


tests/test.php
<?php
/**
 *  test case.
 */
class test extends PHPUnit_Framework_TestCase {
	
	/**
	 * Prepares the environment before running a test.
	 */
	protected function setUp() {
		parent::setUp ();
		$_SERVER['DOCUMENT_ROOT'] = str_replace("tests", "", dirname(__FILE__));
		include_once('/lib/class.detectbrowser.php');
	}
	
	/**
	 * Skyfire test
	 */
	public function testSkyfire()
	{
		// Skyfire
		$detector = new DetectBrowser("Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5_7; en-us) AppleWebKit/530.17 (KHTML, like Gecko) Version/4.0 Safari/530.17 Skyfire/2.0");
		$browser = $detector->get_browser();
		$this->assertEquals('Skyfire', $browser['name']);
	}
	
	/**
	 * Dolphin test
	 */
	public function testDolphin()
	{
		// Skyfire
		$detector = new DetectBrowser("Mozilla/5.0 (SAMSUNG; SAMSUNG-GT-S8500/S8500XXJH3; U; Bada/1.0; en-us) AppleWebKit/533.1 (KHTML, like Gecko) Dolfin/2.0 Mobile WVGA SMM-MMS/1.2.0 OPN-B");
		$browser = $detector->get_browser();
		$this->assertEquals('Dolphin', $browser['name']);
	}
	
	/**
	 * Cleans up the environment after running a test.
	 */
	protected function tearDown() {
		parent::tearDown ();
		unset($_SERVER['DOCUMENT_ROOT']);
	}
	
	/**
	 * Constructs the test case.
	 */
	public function __construct() {
	}

}



tests/testBrowsers.php
<?php
/**
 *  test case.
 */
class testBrowsers extends PHPUnit_Framework_TestCase {
	
	/**
	 * Prepares the environment before running a test.
	 */
	protected function setUp() {
		parent::setUp ();
		$_SERVER['DOCUMENT_ROOT'] = str_replace("tests", "", dirname(__FILE__));
		include_once('/lib/class.detectbrowser.php');
	}
	
	/**
	 * Skyfire test
	 */
	public function testSkyfire()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5_7; en-us) AppleWebKit/530.17 (KHTML, like Gecko) Version/4.0 Safari/530.17 Skyfire/2.0");
		$browser = $detector->getBrowser();
		$this->assertEquals('Skyfire', $browser['name']);
	}
	
	/**
	 * Dolphin test
	 */
	public function testDolphin()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (SAMSUNG; SAMSUNG-GT-S8500/S8500XXJH3; U; Bada/1.0; en-us) AppleWebKit/533.1 (KHTML, like Gecko) Dolfin/2.0 Mobile WVGA SMM-MMS/1.2.0 OPN-B");
		$browser = $detector->getBrowser();
		$this->assertEquals('Dolphin', $browser['name']);
	}
	
	/**
	 * AOL Explorer test
	 */
	public function testAOLExplorer()
	{
		$detector = new DetectBrowser("Mozilla/4.0 (compatible; MSIE 8.0; AOL 9.0; Windows NT 5.1; Trident/4.0; GTB6.5; BO1IE8_v1;ENUS)");
		$browser = $detector->getBrowser();
		$this->assertEquals('AOL Explorer', $browser['name']);
	}
	
	/**
	 * Konqueror test
	 */
	public function testKonqueror()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (compatible; Konqueror/3.1; X11)");
		$browser = $detector->getBrowser();
		$this->assertEquals('Konqueror', $browser['name']);
	}
	
	/**
	 * Netscape Navigator test
	 */
	public function testNetscapeNavigator()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Windows; U; Win 9x 4.90; de-DE; rv:0.9.2) Gecko/20010726 Netscape6/6.1");
		$browser = $detector->getBrowser();
		$this->assertEquals('Netscape Navigator', $browser['name']);
	}
	
	/**
	 * Netscape Navigator test
	 */
	public function testIceWeasel()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (X11; U; Linux i686; en-GB; rv:1.9.0.19) Gecko/2010091807 Iceweasel/3.0.6 (Debian-3.0.6-3)");
		$browser = $detector->getBrowser();
		$this->assertEquals('IceWeasel', $browser['name']);
	}
	
	/**
	 * Beamrise test
	 */
	public function testBeamrise()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US) AppleWebKit/534.13 (KHTML, like Gecko) Chrome/9.0.597.98 Beamrise/4.20.13.22 Safari/534.13");
		$browser = $detector->getBrowser();
		$this->assertEquals('Beamrise', $browser['name']);
	}
	
	/**
	 * Camino test
	 */
	public function  testCamino()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Macintosh; U; PPC Mac OS X Mach-O; en; rv:1.8.1.4pre) Gecko/20070511 Camino/1.6pre");
		$browser = $detector->getBrowser();
		$this->assertEquals('Camino', $browser['name']);
	}
	
	/**
	 * Columbus test
	 */
	public function testColumbus()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Windows; U; Windows NT 6.1; cs-CZ) AppleWebKit/533.3 (KHTML, like Gecko) Columbus/1.2.1.0 Safari/533.3");
		$browser = $detector->getBrowser();
		$this->assertEquals('Columbus', $browser['name']);
	}
	
	/**
	 * Deepnet Exporer test
	 */
	public function testDeepnetExplorer()
	{
		$detector = new DetectBrowser("Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; Deepnet Explorer 1.5.0; .NET CLR 1.0.3705)");
		$browser = $detector->getBrowser();
		$this->assertEquals('Deepnet Explorer', $browser['name']);
	}
	
	/**
	 * CometBird test
	 */
	public function testCometBird()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Windows; U; Windows NT 6.0; en-US; rv:1.9.0.5) Gecko/2009011615 Firefox/3.0.5 CometBird/3.0.5");
		$browser = $detector->getBrowser();
		$this->assertEquals('CometBird', $browser['name']);
	}
	
	/**
	 * Kylo test
	 */
	
	public function testKylo()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.9.2) Gecko/20100222 Firefox/3.6 Kylo/0.8.4.74873");
		$browser = $detector->getBrowser();
		$this->assertEquals('Kylo', $browser['name']);
	}
	
	/**
	 * iCab test
	 */
	
	public function testICab()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Macintosh; PPC Mac OS X 10_5_8) AppleWebKit/537.3+ (KHTML, like Gecko) iCab/5.0 Safari/533.16");
		$browser = $detector->getBrowser();
		$this->assertEquals('iCab', $browser['name']);
	}
	
	public function testChimera()
	{
		$detector = new DetectBrowser("Mozilla/5.0 (Macintosh; U; PPC Mac OS X; pl-PL; rv:1.0.1) Gecko/20021111 Chimera/0.6");
		$browser = $detector->getBrowser();
		$this->assertEquals('Chimera', $browser['name']);
	}
	
	/**
	 * Cleans up the environment after running a test.
	 */
	protected function tearDown() {
		parent::tearDown ();
		unset($_SERVER['DOCUMENT_ROOT']);
	}
		
	/**
	 * Constructs the test case.
	 */
	public function __construct() {
	}

}



example.php
<?php
include_once($_SERVER['DOCUMENT_ROOT'].'/lib/class.detectbrowser.php');
$endorphin = new DetectBrowser("Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_5_7; en-us) AppleWebKit/530.17 (KHTML, like Gecko) Version/4.0 Safari/530.17 Skyfire/2.0");
$browser = $endorphin->getBrowser();
$os = $endorphin->getOs();
$device = $endorphin->getDevice();
?>
<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8" />
		<title>Endorphin PHP lib Example</title>
		<style>
			html
			{
				background: url("/images/bg.png");
				font-size: 12px;
				font-family: Segoe UI,Ubuntu,Tahoma;
			}
			.container
			{
				width: 400px;
				margin: 10% auto;
			}
			.block
			{
				background-color: #000;
				overflow: auto;
				opacity: 0.7;
				-webkit-border-radius: 10px;
				-moz-border-radius: 10px;
				border-radius: 10px;
				margin: 10px 0px;
			}
			.block *
			{
				opacity: 1;
			}

			.icon
			{
				width: 96px;
				height: 96px;
				padding: 15px;
				float: left;
			}
			.info
			{
				float: left;
				vertical-align: center;
				padding: 30px 25px;
				color: #fff;
				height: 100%;
				margin: auto 20px;
				width: 170px;
				min-height: 40px;
			}
		</style>
	</head>
	<body>
		<div class="container">
			<div class="block">
				<div class="icon">
					<img src="/images/<? echo $browser['name'];?>.png" />
				</div>
				<div class="info">
					<div>Your Browser is: <? echo $browser['name']; ?></div>
					<div>Browser version: <? echo $browser['version']; ?></div>
					<div>Browser type: <? echo $browser['type']; ?></div>
				</div>
			</div>
			<div class="block">
				<div class="icon">
					<img src="/images/<? echo $device['type'];?>.png" />
				</div>
				<div class="info">
					<div>Your Device is: <? echo $device['name']; ?></div>
					<div>Device type: <? echo $device['type']; ?></div>
				</div>
			</div>
			<div class="block">
				<div class="icon">
					<img src="/images/<? echo $os['name'];?>.png" />
				</div>
				<div class="info">
					<div>Your OS is: <? echo $os['name']; ?></div>
					<? if(isset($os['version']))
					{?>
					<div>OS version: <? echo $os['version']; ?></div>
					<?}?>
					<div>OS type: <? echo $os['type']; ?></div>
					<div>OS family: <? echo $os['family']; ?></div>
				</div>
			</div>
		</div>
	</body>
</html>


lib/class.detectbrowser.php
<?php
/**
 * @version 1.1.9
 * @author Sergey Nehaenko <sergey.nekhaenko@gmail.com>
 * @copyright Copyright (c) Sergey Nehaenko 2012-2015
 * @package ru.endorphinua.detector
 * @license: GPL
 */
define('ROOT',$_SERVER['DOCUMENT_ROOT']);
class DetectBrowser
{
	/**
	 * @var array contains information about the user's operating system
	 */
	private $os = Array('name'=>'none','family'=>'none');
	/**
	 * @var array contains information about the device on which the site is viewed
	 */
	private $device = Array('name'=>'none','type'=>'none'); 
	/**
	 * @var array containing information about the user's browser
	 */
	private $browser = Array('name'=>'none','version'=>'none','type'=>'none');
	/**
	 * @var string User Agent string
	 */
	private $ua; // user agent
	/**
	 * @var array contains the path to the xml files
	 */
	private $xml_path = Array('os'=> '/data/os.xml','browser'=> '/data/browsers.xml','device'=>'/data/device.xml','bot'=>'/data/bot.xml');
	/**
	 * @var array contains information from xml files
	 */
	private $xml;
	
	public $isBot = false;
	
	private $debug = false;
	
	/*
	 * Оbject initialization
	 * @param string $ua User Agent string
	 */
	public function __construct($ua = 'none',$debug=false)
	{
		if($ua == 'none')
		{
			$this->ua = $_SERVER['HTTP_USER_AGENT'];
		}
		else
		{
			$this->ua = $ua;
		}
		$this->debug = $debug;
		$this->loadXml();
		$this->detectBot();
		if(!$this->isBot)
		{
			$this->detectBroswer();
			$this->detectDevice();
			$this->detectOs();
			$this->fix();
		}
		
		if($this->debug)
		{
			$this->printAll();
		}
	}
	
	/**
	 * Load the contents of xml files into an array
	 */
	private function loadXml()
	{
		$os = simplexml_load_file(ROOT.$this->xml_path['os']);
		$browser = simplexml_load_file(ROOT.$this->xml_path['browser']);
		$device = simplexml_load_file(ROOT.$this->xml_path['device']);
		$bot = simplexml_load_file(ROOT.$this->xml_path['bot']);
		
		$this->xml = Array();
		foreach($os->children() as $element)
		{
			$id = $element->attributes()->id;
			foreach($element->children() as $property_key => $property_value)
			{
				$property = (string) $property_key;
				$id = (string) $id;
				$this->xml['os'][$property][$id] = (string)$property_value;
			}
		}
		foreach($browser->children() as $element)
		{
			$id = $element->attributes()->id;
			foreach($element->children() as $property_key => $property_value)
			{
				$property = (string) $property_key;
				$id = (string) $id;
				$this->xml['browser'][$property][$id] = (string)$property_value;
			}
		}
		foreach($device->children() as $element)
		{
			$id = $element->attributes()->id;
			foreach($element->children() as $property_key => $property_value)
			{
				$property = (string) $property_key;
				$id = (string) $id;
				$this->xml['device'][$property][$id] = (string)$property_value;
			}
		}
		foreach($bot->children() as $element)
		{
			$id = $element->attributes()->id;
			foreach($element->children() as $property_key => $property_value)
			{
				$property = (string) $property_key;
				$id = (string) $id;
				$this->xml['bot'][$property][$id] = (string)$property_value;
			}
		}
	}
	
	/**
	 * Getter for BROWSER property
	*/
	
	public function getBrowser()
	{
		return $this->browser;
	}
	
	/**
	 * Getter for OS property
	*/ 

	public function getOs()
	{
		return $this->os;
	}

	/**
	 * Getter for DEVICE property
	*/ 

	public function getDevice()
	{
		return $this->device;
	}
	
	/**
	 * Getter for USER AGENT property
	 */ 
	
	public function getUa()
	{
		return $this->ua;
	}
	
	/**
	 * Detect information about BROWSER
	*/ 
	
	private function detectBroswer()
	{	
		$pattern = $this->xml['browser']['pattern'];
		
		foreach($pattern as $key=>$value)
		{
			$value = '/'.$value.'/';
			if(preg_match($value,$this->ua))
			{
				if(isset($this->xml['browser']['version_pattern'][$key]))
				{
					$this->setBrowser($key,$this->xml['browser']['type'][$key],$this->xml['browser']['version_pattern'][$key]);
				}
				else
				{
					$this->setBrowser($key,$this->xml['browser']['type'][$key]);
				}
				break;
			}
		}
	}
	
	/**
	 * Setter for BROWSER property
	*/ 
	
	private function setBrowser($name,$type,$phrase='same')
	{
		$this->browser['type'] = $type;
		$this->browser['name'] = $name;
		if($phrase == 'same')
		{
			$this->detectBrowserVersion($name);
		}
		else
		{
			$this->detectBrowserVersion($phrase);
		}
		if($name == 'Opera Mini')
		{
			$this->ua = $_SERVER['HTTP_X_OPERAMINI_PHONE_UA'];
		}
	}
	
	/**
	 * Detect BROWSER version
	*/ 
	 
	private function detectBrowserVersion($phrase)
	{
		$version = "/".$phrase."(\/| )[\w-._]{1,15}/";
		if(preg_match($version,$this->ua))
		{
			preg_match($version,$this->ua,$v);
			@$version = $v[0];
			$version = preg_replace('/'.$phrase.'/','',$version);
			$version = str_replace(';','',$version);
			$version = str_replace(' ','',$version);
			$version = str_replace('/','',$version);
			$version = str_replace('_','.',$version);
			$this->browser['version'] = $version;
		}
	}
	
	/**
	 * Detect information about DEVICE
	*/ 
	
	private function detectDevice()
	{
		$pattern = $this->xml['device']['pattern'];
		
		foreach($pattern as $key=>$value)
		{
			$value = '/'.$value.'/';
			if(preg_match($value,$this->ua))
			{
				if($key == 'Lumia')
				{
					preg_match($value,$this->ua,$r);
					$this->setDevice($r[0],$this->xml['device']['type'][$key]);
				}
				else
				{
					$this->setDevice($key,$this->xml['device']['type'][$key]);
				}
				break;
			}
		}
	}
	
	/**
	 * Setter of DEVICE property
	*/ 
	
	private function setDevice($name,$type)
	{
		$this->device['name'] = $name;
		$this->device['type'] = $type;
	}
	
	/*
	 * Detect information about OS
	*/ 
	private function detectOs()
	{		
		$pattern = $this->xml['os']['pattern'];
		
		foreach($pattern as $key=>$value)
		{
			$value = '/'.$value.'/';
			if(preg_match($value,$this->ua))
			{
				if(isset($this->xml['os']['version_pattern'][$key]))
				{
					$this->setOs($key,$this->xml['os']['family'][$key],$this->xml['os']['version_pattern'][$key]);
				}
				else
				{
					$this->setOs($key,$this->xml['os']['family'][$key]);
				}
				break;
			}
		}
	}
	
	/*
	 * Detect sitebot
	*/ 
	private function detectBot()
	{		
		$pattern = $this->xml['bot']['pattern'];
		
		foreach($pattern as $key=>$value)
		{
			$value = '/'.$value.'/';
			if(preg_match($value,$this->ua))
			{
				$this->isBot = true;
				$this->setDevice($key, $this->xml['bot']['device'][$key]);
				break;
			}
		}
	}
	
	/*
	 * Setter for OS property
	*/ 
	
	private function setOs($name,$family,$phrase='same')
	{
		$this->os['name'] = $name;
		$this->os['family'] = $family;
		$this->os['type'] = $this->xml['os']['type'][$name];
		if($phrase != 'same')
		{
			$this->detectOsVersion($phrase);
		}
	}
	
	/**
	 * Detect OS version
	*/ 
	
	private function detectOsVersion($phrase)
	{
		if($this->os['family']!='win')
		{
			$version = "/".$phrase."(\/| )?[\w-._]{1,15}/";
			$reverse = "/[\w-._]{1,15}(\/| )?".$phrase."/";
			if(preg_match($version,$this->ua))
			{
				preg_match($version,$this->ua,$v);
				@$version = $v[0];
				$phrase = '/'.$phrase.'/';
				$version = preg_replace($phrase,'',$version);
				$version = str_replace(';','',$version);
				$version = str_replace(' ','',$version);
				$version = str_replace('/','',$version);
				$version = str_replace('_','.',$version);
				$this->os['version'] = $version;
			}
			else
			if(preg_match($reverse,$this->ua))
			{
				preg_match($reverse,$this->ua,$v);
				@$version = $v[0];
				$phrase = '/'.$phrase.'/';
				$version = preg_replace($phrase,'',$version);
				$version = str_replace(';','',$version);
				$version = str_replace(' ','',$version);
				$version = str_replace('/','',$version);
				$version = str_replace('_','.',$version);
				$this->os['version'] = $version;
			}
		}
		else
		{
			$pattern['95'] = '/(Win95)/';
			$pattern['98'] = '/(Win98)/';
			$pattern['ME'] = '/(Win 9[0-9]{1} 4.90|WinNT|Windows ME)/';
			$pattern['2000'] = '/(Windows NT 5.0|Windows 2000|Windows NT 5.01)/';
			$pattern['XP'] = '/Windows NT 5.1/';
			$pattern['Server 2003'] = '/Windows NT 5.2/';
			$pattern['Vista'] = '/Windows NT 6.0/';
			$pattern['7'] = '/Windows NT 6.1/';
			$pattern['8'] = '/Windows NT 6.2/';
			$pattern['CE'] = '/Windows CE/';
			$pattern['Phone'] = '/(Windows Phone OS|WP|Windows Phone|Windows\+Phone|Windows\+Phone\+OS)/';
			$pattern['Mobile'] = '/Windows Mobile/';
			
			foreach($pattern as $key=>$value)
			{
				if(preg_match($value,$this->ua))
				{
					$this->os['version'] = $key;
					break;
				}
			}
			if($this->os['version'] == 'Phone')
			{
				$this->detectWindowsPhoneVersion();
			}
		}
		$this->fixOsVersion();
	}
	
	/**
	 * Detection of Windows Phone Version;
	 */
	private function detectWindowsPhoneVersion()
	{
		$pattern = '/(Windows Phone OS|Windows Phone|Windows\+Phone|Windows\+Phone\+OS) [\w-._\+]{1,15}/';
		if(preg_match($pattern, $this->ua))
		{
			preg_match($pattern, $this->ua,$v);
			$v = $v[0];
			$v = str_replace('+', ' ', $v);
			$v = str_replace('Windows Phone OS', '', $v);
			$v = str_replace('Windows Phone', '', $v);
			$v = str_replace('Windows Phone', '', $v);
			$v = str_replace(' ', '', $v);
			$v = str_replace('\\', '', $v);
			$v = str_replace('|', '', $v);
			$this->os['name'] .= ' '.$this->os['version'];
			$this->os['version'] = ' '.$v;
		}
		$this->os['type'] == 'mobile';
	}
	
	/**
	 * Fix problems with detection of OS version
	 */
	private function fixOsVersion()
	{
		if(isset($this->os['version']))
		{
			$replace = array('based.','Fir','fc','update');
			foreach($replace as $key)
			{
				$this->os['version'] = str_replace($key,'',$this->os['version']);
			}
		}
	}
	
	/**
	 * Fix some problems
	*/
	private function fix()
	{
		/* fix iPhone detection */
		if($this->device['name'] == 'iPhone' && $this->browser['name'] == 'none')
		{
			$this->browser['name'] = 'Safari';
			$this->detectBrowserVersion('Version');
		}
		
		/* Fix Chrome Mobile Detection */
		if($this->browser['name'] == 'Google Chrome' && $this->os['name'] == 'Android' || $this->browser['name'] == 'Google Chrome' && $this->os['name'] == 'iOS' || $this->browser['name'] == 'Google Chrome' && $this->device['type'] == 'mobile' || $this->browser['name'] == 'Google Chrome' && $this->device['type'] == 'tablet')
		{
			$this->browser['name'] = 'Chrome Mobile';
			$this->browser['type'] = 'mobile';
		}
		
		/* Fix Mobile Safari detection */
		if($this->device['type'] == 'mobile' && $this->browser['name'] == 'Safari' || $this->device['type'] == 'tablet' && $this->browser['name'] == 'Safari')
		{
			$this->browser['type'] = 'mobile';
		} 
		
		/* Fix Mobile Firefox detection */
		if($this->browser['name'] == 'Firefox' && $this->device['type'] == 'mobile' || $this->browser['name'] == 'Firefox' && $this->os['name'] == 'Android')
		{
			$this->browser['type'] = 'mobile';
		}
		
		/* Fix Chrome mobile detection */
		if($this->browser['name'] == 'Google Chrome' && $this->os['name'] == 'Android' || $this->browser['name'] == 'Google Chrome' && $this->os['name'] == 'iOS')
		{
			$this->browser['type'] = 'mobile';
		}
		
		/* Fix Android browser detection */
		if($this->os['name'] == 'Android' && $this->browser['name'] == 'Safari')
		{
			$this->browser['name'] = 'Android Browser';
		}
		
		/* Fix PC detection */
		if($this->os['family'] == 'win' || $this->os['name'] == 'Mac OS X' || $this->browser['type'] == 'desktop')
		{
			if($this->device['type'] == 'none')
			{
				$this->setDevice('PC','pc');
			}
		}
		
		/* Fix Windows Phone os['type'] bug */
		if($this->os['family'] == 'win' && $this->device['type'] == 'mobile')
		{
			$this->os['type'] = 'mobile';
		}
		
		/* Fix Google phone detection */
		if($this->device['name'] == 'Google' && $this->os['name'] == 'Android')
		{
			$this->device['name'] = 'Android';
		}
		
		/* Fix Android mobile phones detection */
		if($this->os['name'] == 'Android' && preg_match('/Mobile/',$this->ua))
		{
			if($this->device['name'] == 'none')
			{
				$this->device['name'] = 'Android';
			}
			$this->device['type'] = 'mobile';
		}
		/* Fix Android tablets detection */
		if($this->os['name'] == 'Android' && !preg_match('/Mobile/',$this->ua))
		{
			if($this->device['name'] == 'none')
			{
				$this->device['name'] = 'Android';
			}
			$this->device['type'] = 'tablet';
		}
		
		/* Fix Mobile phones detection */
		
		if($this->browser['name'] == 'Opera Mini' || $this->browser['name'] == 'Opera Mobile')
		{
			if($this->device['type'] == 'none')
			{
				$this->setDevice('Mobile phone','mobile');
			}
		}
		
		/* Last MICROSOFT LUMIAs FIX */
		if($this->browser['name'] == 'none' && $this->device['type'] == 'mobile' && $this->os['name'] = 'Windows')
		{
			$this->setBrowser('Internet Explorer Mobile','mobile');
		}
	}
	
	public function printAll()
	{
		echo $this->getUa()."\n";
		foreach ($this->getAll() as $key=>$value)
		{
			echo $key."\n";
			foreach ($value as $key2=>$val)
			{
				echo "\t".$key2.': '.$val."\n";
			}
		}
		echo "\n";
	}
	
	public function getAll()
	{
		return Array('browser'=>$this->browser,'os'=>$this->os,'device'=>$this->device);
	}
}
?>
